File: results/2026-07-01_184135/tasks/17-artifact-cleanup-script/default-fable5-high/generated-code/run_act_tests.py
"""End-to-end test harness: every test case runs through the GitHub Actions
workflow via `act` (nektos/act), never against the script directly.

For each case it:
  1. builds a temp git repo containing the project files plus that case's
     fixture data (artifacts inventory, retention policy, pinned "now"),
  2. runs `act push --rm` in it,
  3. appends the full act output to act-result.txt (clearly delimited),
  4. asserts act exited 0, that BOTH jobs report "Job succeeded", and that
     the output contains the exact known-good values for that input.

Workflow structure tests (YAML shape + actionlint) run first, before any
act time is spent.

Usage: python3 run_act_tests.py
"""

import os
import shutil
import subprocess
import sys
import tempfile

PROJECT_DIR = os.path.dirname(os.path.abspath(__file__))
RESULT_FILE = os.path.join(PROJECT_DIR, "act-result.txt")
# Files/dirs the workflow needs inside each temp repo.
PROJECT_FILES = ["artifact_cleanup.py", "tests", ".github"]
ACT_CMD = [
    "act", "push", "--rm",
    "-P", "ubuntu-latest=act-ubuntu-pwsh:latest",
    "--pull=false",
    "--action-offline-mode",
]
JOB_NAMES = ["Unit tests", "Cleanup plan"]

# ---------------------------------------------------------------------------
# Test cases. `expected` substrings are EXACT known-good values, derived by
# hand from each case's inputs (see comments), not from running the script.
# ---------------------------------------------------------------------------
CASES = [
    {
        # All three policies fire at once (reference time 2026-07-01):
        #  - old-build-logs (2026-05-01) is 61 days old   -> max-age (>30)
        #  - run 100 has 3 artifacts, keep_latest_n=2     -> rank 3 deleted
        #  - survivors total 14 MB against a 10 MB cap    -> evict oldest (4 MB)
        # Deleted: 2 MB + 3 MB + 4 MB = 9 MB reclaimed; 3 retained (10 MB).
        "name": "combined-policies",
        "now": "2026-07-01T00:00:00Z",
        "artifacts": [
            {"name": "old-build-logs", "size_bytes": 2097152,
             "created_at": "2026-05-01T00:00:00Z", "workflow_run_id": 99},
            {"name": "run1-artifact-a", "size_bytes": 3145728,
             "created_at": "2026-06-10T00:00:00Z", "workflow_run_id": 100},
            {"name": "run1-artifact-b", "size_bytes": 4194304,
             "created_at": "2026-06-15T00:00:00Z", "workflow_run_id": 100},
            {"name": "run1-artifact-c", "size_bytes": 1048576,
             "created_at": "2026-06-20T00:00:00Z", "workflow_run_id": 100},
            {"name": "run2-artifact-a", "size_bytes": 5242880,
             "created_at": "2026-06-25T00:00:00Z", "workflow_run_id": 200},
            {"name": "run2-artifact-b", "size_bytes": 4194304,
             "created_at": "2026-06-28T00:00:00Z", "workflow_run_id": 200},
        ],
        "policy": {"max_age_days": 30, "keep_latest_n": 2,
                   "max_total_size_bytes": 10485760},
        "expected": [
            "DELETE old-build-logs (2.00 MB) - max-age: 61 days old exceeds limit of 30 days",
            "DELETE run1-artifact-a (3.00 MB) - keep-latest: rank 3 of 3 in workflow run 100 exceeds keep-latest limit of 2",
            "DELETE run1-artifact-b (4.00 MB) - max-total-size: retained total 14.00 MB exceeds limit of 10.00 MB",
            "KEEP   run1-artifact-c (1.00 MB)",
            "KEEP   run2-artifact-a (5.00 MB)",
            "KEEP   run2-artifact-b (4.00 MB)",
            "Total artifacts: 6",
            "Retained:        3",
            "Deleted:         3",
            "Space reclaimed: 9.00 MB",
            "Space retained:  10.00 MB",
            "Dry run: 3 artifact(s) would be deleted, reclaiming 9.00 MB",
            # apply step (oldest first):
            "Deleting artifact 'old-build-logs' (workflow run 99)... deleted",
            "Deleting artifact 'run1-artifact-a' (workflow run 100)... deleted",
            "Deleting artifact 'run1-artifact-b' (workflow run 100)... deleted",
            "Deleted 3 artifact(s), reclaimed 9.00 MB",
        ],
    },
    {
        # Nothing violates any policy: everything is retained, both the
        # dry-run and the apply step must be no-ops.
        "name": "nothing-to-delete",
        "now": "2026-07-01T00:00:00Z",
        "artifacts": [
            {"name": "nightly-report", "size_bytes": 1048576,
             "created_at": "2026-06-29T00:00:00Z", "workflow_run_id": 300},
            {"name": "coverage-html", "size_bytes": 2097152,
             "created_at": "2026-06-30T00:00:00Z", "workflow_run_id": 301},
        ],
        "policy": {"max_age_days": 90, "keep_latest_n": 10,
                   "max_total_size_bytes": 1073741824},
        "expected": [
            "KEEP   nightly-report (1.00 MB)",
            "KEEP   coverage-html (2.00 MB)",
            "Total artifacts: 2",
            "Retained:        2",
            "Deleted:         0",
            "Space reclaimed: 0 B",
            "Space retained:  3.00 MB",
            "Dry run: 0 artifact(s) would be deleted, reclaiming 0 B",
            "No artifacts to delete.",
        ],
    },
    {
        # Age-only policy (the other rules are omitted, i.e. disabled):
        # stale-cache (2026-01-01) is 181 days old -> deleted, 5 MB reclaimed.
        "name": "age-only-policy",
        "now": "2026-07-01T00:00:00Z",
        "artifacts": [
            {"name": "stale-cache", "size_bytes": 5242880,
             "created_at": "2026-01-01T00:00:00Z", "workflow_run_id": 1},
            {"name": "fresh-build", "size_bytes": 1048576,
             "created_at": "2026-06-30T00:00:00Z", "workflow_run_id": 2},
        ],
        "policy": {"max_age_days": 30},
        "expected": [
            "DELETE stale-cache (5.00 MB) - max-age: 181 days old exceeds limit of 30 days",
            "KEEP   fresh-build (1.00 MB)",
            "Total artifacts: 2",
            "Retained:        1",
            "Deleted:         1",
            "Space reclaimed: 5.00 MB",
            "Space retained:  1.00 MB",
            "Dry run: 1 artifact(s) would be deleted, reclaiming 5.00 MB",
            "Deleting artifact 'stale-cache' (workflow run 1)... deleted",
            "Deleted 1 artifact(s), reclaimed 5.00 MB",
        ],
    },
]

failures = []


def check(case_name, condition, message):
    """Record (and print) one assertion result."""
    status = "PASS" if condition else "FAIL"
    print(f"  [{status}] {message}")
    if not condition:
        failures.append(f"{case_name}: {message}")


def build_temp_repo(case, root):
    """Copy the project + this case's fixtures into a fresh git repo."""
    import json
    for name in PROJECT_FILES:
        src = os.path.join(PROJECT_DIR, name)
        dst = os.path.join(root, name)
        if os.path.isdir(src):
            shutil.copytree(src, dst)
        else:
            shutil.copy2(src, dst)
    fixtures = os.path.join(root, "fixtures")
    os.makedirs(fixtures)
    with open(os.path.join(fixtures, "artifacts.json"), "w") as fh:
        json.dump(case["artifacts"], fh, indent=2)
    with open(os.path.join(fixtures, "policy.json"), "w") as fh:
        json.dump(case["policy"], fh, indent=2)
    with open(os.path.join(fixtures, "now.txt"), "w") as fh:
        fh.write(case["now"] + "\n")

    env = dict(os.environ,
               GIT_AUTHOR_NAME="harness", GIT_AUTHOR_EMAIL="harness@test",
               GIT_COMMITTER_NAME="harness", GIT_COMMITTER_EMAIL="harness@test")
    for cmd in (["git", "init", "-q"],
                ["git", "add", "-A"],
                ["git", "commit", "-q", "-m", f"fixture: {case['name']}"]):
        subprocess.run(cmd, cwd=root, env=env, check=True,
                       capture_output=True, text=True)


def run_structure_tests():
    """Workflow structure tests (YAML shape, file refs, actionlint)."""
    print("== Workflow structure tests ==")
    result = subprocess.run(
        [sys.executable, "-m", "unittest", "test_workflow_structure", "-v"],
        cwd=PROJECT_DIR, capture_output=True, text=True)
    print(result.stderr.strip().splitlines()[-1])
    check("structure", result.returncode == 0,
          "workflow structure tests (incl. actionlint) exit 0")
    if result.returncode != 0:
        print(result.stdout)
        print(result.stderr)
    return result


def run_case(index, case, result_fh):
    print(f"\n== Case {index}: {case['name']} ==")
    with tempfile.TemporaryDirectory(prefix=f"act-{case['name']}-") as root:
        build_temp_repo(case, root)
        proc = subprocess.run(ACT_CMD, cwd=root, capture_output=True,
                              text=True, timeout=900)
    output = proc.stdout + proc.stderr

    result_fh.write(f"\n{'=' * 78}\n")
    result_fh.write(f"=== CASE {index}: {case['name']} "
                    f"(act exit code {proc.returncode}) ===\n")
    result_fh.write(f"{'=' * 78}\n")
    result_fh.write(output)
    result_fh.flush()

    check(case["name"], proc.returncode == 0, "act exited with code 0")
    for job in JOB_NAMES:
        succeeded = any(
            job in line and "Job succeeded" in line
            for line in output.splitlines())
        check(case["name"], succeeded, f"job '{job}' reports 'Job succeeded'")
    check(case["name"], "Job failed" not in output, "no job reports 'Job failed'")
    for fragment in case["expected"]:
        check(case["name"], fragment in output,
              f"output contains exact value: {fragment!r}")


def main():
    with open(RESULT_FILE, "w") as result_fh:
        result_fh.write("act end-to-end test results\n")
        run_structure_tests()
        for index, case in enumerate(CASES, start=1):
            run_case(index, case, result_fh)
        result_fh.write(f"\n{'=' * 78}\n")
        verdict = "ALL ASSERTIONS PASSED" if not failures else \
            f"{len(failures)} ASSERTION(S) FAILED"
        result_fh.write(f"HARNESS VERDICT: {verdict}\n")
        for failure in failures:
            result_fh.write(f"  FAIL {failure}\n")

    print(f"\n== Verdict: {verdict} (full act output in act-result.txt) ==")
    return 1 if failures else 0


if __name__ == "__main__":
    sys.exit(main())

File: results/2026-07-01_184135/tasks/17-artifact-cleanup-script/typescript-bun-fable5-high/generated-code/scripts/run-act-tests.ts
/**
 * act-based end-to-end test harness.
 *
 * For each test case this script:
 *   1. Creates a temp git repo containing the full project plus that case's
 *      fixture data (fixtures/artifacts.json + fixtures/policy.json),
 *   2. Runs `act push --rm` against it (executing the real workflow:
 *      unit tests job -> cleanup-plan job),
 *   3. Appends the full act output to act-result.txt (clearly delimited),
 *   4. Asserts act exited 0, both jobs report "Job succeeded", and the
 *      ::PLAN::{json}::ENDPLAN:: line parsed from the CI logs matches the
 *      known-good expected values EXACTLY (summary numbers, artifact names,
 *      per-artifact deletion reasons, dry-run flag).
 *
 * Run with: bun run scripts/run-act-tests.ts
 * (Deliberately NOT a *.test.ts file: the workflow itself runs `bun test`,
 * and this harness must not recurse inside the act container.)
 */
import { cpSync, mkdtempSync, rmSync } from "node:fs";
import { tmpdir } from "node:os";
import { join } from "node:path";
import type { Artifact, PlanSummary } from "../src/types";

const ROOT = join(import.meta.dir, "..");
const RESULT_FILE = join(ROOT, "act-result.txt");

/** Files copied into each temp repo. node_modules is included so
 * `bun install --frozen-lockfile` inside the container is a no-op-fast path. */
const PROJECT_FILES: string[] = [
  ".github",
  ".actrc",
  "src",
  "tests",
  "fixtures",
  "scripts",
  "node_modules",
  "package.json",
  "tsconfig.json",
  "bun.lock",
];

interface ExpectedOutcome {
  summary: PlanSummary;
  deletedNames: string[];
  retainedNames: string[];
  dryRun: boolean;
  /** name -> exact reasons list, for spot-checking policy attribution. */
  reasons: Record<string, string[]>;
  stdoutContains: string[];
  stdoutNotContains: string[];
}

interface ActCase {
  name: string;
  artifacts: Artifact[];
  policy: Record<string, unknown>;
  expected: ExpectedOutcome;
}

// ---------------------------------------------------------------------------
// Test cases with hand-computed known-good outcomes.
// ---------------------------------------------------------------------------
const CASES: ActCase[] = [
  {
    // All three policies at once, dry-run. Reference date 2026-07-01:
    // A1 is >30d old AND ranked 4th of 4 in run 100 -> max-age + keep-latest.
    // A2 is ranked 3rd of 4 in run 100 -> keep-latest.
    // Survivors A3+A4+A5+A6 = 410 bytes > 250 cap -> evict oldest (A5, 200).
    name: "combined-policies-dry-run",
    artifacts: [
      { id: 1, name: "A1", sizeBytes: 100, createdAt: "2026-05-01T00:00:00Z", workflowRunId: 100 },
      { id: 2, name: "A2", sizeBytes: 50, createdAt: "2026-06-20T00:00:00Z", workflowRunId: 100 },
      { id: 3, name: "A3", sizeBytes: 60, createdAt: "2026-06-25T00:00:00Z", workflowRunId: 100 },
      { id: 4, name: "A4", sizeBytes: 70, createdAt: "2026-06-28T00:00:00Z", workflowRunId: 100 },
      { id: 5, name: "A5", sizeBytes: 200, createdAt: "2026-06-10T00:00:00Z", workflowRunId: 200 },
      { id: 6, name: "A6", sizeBytes: 80, createdAt: "2026-06-29T00:00:00Z", workflowRunId: 200 },
    ],
    policy: {
      maxAgeDays: 30,
      keepLatestPerWorkflow: 2,
      maxTotalSizeBytes: 250,
      referenceDate: "2026-07-01T00:00:00Z",
      dryRun: true,
    },
    expected: {
      summary: {
        totalArtifacts: 6,
        retainedCount: 3,
        deletedCount: 3,
        spaceReclaimedBytes: 350,
        retainedSizeBytes: 210,
      },
      deletedNames: ["A1", "A2", "A5"],
      retainedNames: ["A3", "A4", "A6"],
      dryRun: true,
      reasons: {
        A1: ["max-age", "keep-latest-per-workflow"],
        A2: ["keep-latest-per-workflow"],
        A5: ["max-total-size"],
      },
      stdoutContains: [
        "Artifact Cleanup Plan (DRY RUN)",
        "Summary: 6 total | 3 retained (210 bytes) | 3 to delete | 350 bytes reclaimed",
        "Dry run: 3 artifacts would be deleted, nothing was touched",
      ],
      stdoutNotContains: ["DELETED "],
    },
  },
  {
    // max-age only, EXECUTE mode. "edge" is exactly 7 days old -> retained
    // (only strictly older artifacts are deleted).
    name: "max-age-execute",
    artifacts: [
      { id: 1, name: "old-a", sizeBytes: 10, createdAt: "2026-06-01T00:00:00Z", workflowRunId: 1 },
      { id: 2, name: "old-b", sizeBytes: 20, createdAt: "2026-06-20T00:00:00Z", workflowRunId: 1 },
      { id: 3, name: "new-c", sizeBytes: 30, createdAt: "2026-06-30T00:00:00Z", workflowRunId: 2 },
      { id: 4, name: "edge", sizeBytes: 5, createdAt: "2026-06-24T00:00:00Z", workflowRunId: 3 },
    ],
    policy: { maxAgeDays: 7, referenceDate: "2026-07-01T00:00:00Z", dryRun: false },
    expected: {
      summary: {
        totalArtifacts: 4,
        retainedCount: 2,
        deletedCount: 2,
        spaceReclaimedBytes: 30,
        retainedSizeBytes: 35,
      },
      deletedNames: ["old-a", "old-b"],
      retainedNames: ["edge", "new-c"],
      dryRun: false,
      reasons: { "old-a": ["max-age"], "old-b": ["max-age"] },
      stdoutContains: [
        "Summary: 4 total | 2 retained (35 bytes) | 2 to delete | 30 bytes reclaimed",
        "DELETED old-a (id=1, 10 bytes)",
        "DELETED old-b (id=2, 20 bytes)",
        "Deleted 2 artifacts, reclaimed 30 bytes",
      ],
      stdoutNotContains: ["DRY RUN"],
    },
  },
  {
    // keep-latest-1 + size cap, dry-run. r1-old loses the rank race in run 10;
    // survivors r1-new(90)+r2-only(50)=140 > 100 cap -> evict oldest survivor
    // (r2-only) -> 90 <= 100.
    name: "keep-latest-and-size-cap-dry-run",
    artifacts: [
      { id: 1, name: "r1-old", sizeBytes: 40, createdAt: "2026-06-01T00:00:00Z", workflowRunId: 10 },
      { id: 2, name: "r1-new", sizeBytes: 90, createdAt: "2026-06-15T00:00:00Z", workflowRunId: 10 },
      { id: 3, name: "r2-only", sizeBytes: 50, createdAt: "2026-06-10T00:00:00Z", workflowRunId: 20 },
    ],
    policy: {
      keepLatestPerWorkflow: 1,
      maxTotalSizeBytes: 100,
      referenceDate: "2026-07-01T00:00:00Z",
      dryRun: true,
    },
    expected: {
      summary: {
        totalArtifacts: 3,
        retainedCount: 1,
        deletedCount: 2,
        spaceReclaimedBytes: 90,
        retainedSizeBytes: 90,
      },
      deletedNames: ["r1-old", "r2-only"],
      retainedNames: ["r1-new"],
      dryRun: true,
      reasons: { "r1-old": ["keep-latest-per-workflow"], "r2-only": ["max-total-size"] },
      stdoutContains: [
        "Artifact Cleanup Plan (DRY RUN)",
        "Summary: 3 total | 1 retained (90 bytes) | 2 to delete | 90 bytes reclaimed",
      ],
      stdoutNotContains: ["DELETED "],
    },
  },
];

// ---------------------------------------------------------------------------
// Tiny assertion helpers (failures are collected per case, not thrown midway,
// so act-result.txt is always fully written).
// ---------------------------------------------------------------------------
const failures: string[] = [];

function check(caseName: string, what: string, ok: boolean, detail = ""): void {
  const status = ok ? "PASS" : "FAIL";
  console.log(`  [${status}] ${what}${ok || !detail ? "" : ` — ${detail}`}`);
  if (!ok) failures.push(`${caseName}: ${what}${detail ? ` — ${detail}` : ""}`);
}

function checkEqual(caseName: string, what: string, actual: unknown, expected: unknown): void {
  const a = JSON.stringify(actual);
  const e = JSON.stringify(expected);
  check(caseName, what, a === e, `expected ${e}, got ${a}`);
}

function run(cmd: string[], cwd: string): { exitCode: number; output: string } {
  const proc = Bun.spawnSync(cmd, { cwd, stdout: "pipe", stderr: "pipe" });
  const output = proc.stdout.toString() + proc.stderr.toString();
  return { exitCode: proc.exitCode ?? 1, output };
}

// ---------------------------------------------------------------------------
// Main
// ---------------------------------------------------------------------------
let resultLog = "";

for (const testCase of CASES) {
  console.log(`\n=== act case: ${testCase.name} ===`);
  const tempRepo = mkdtempSync(join(tmpdir(), `artifact-cleanup-${testCase.name}-`));

  try {
    // 1. Assemble the temp repo: project files + case-specific fixtures.
    for (const entry of PROJECT_FILES) {
      cpSync(join(ROOT, entry), join(tempRepo, entry), { recursive: true });
    }
    await Bun.write(
      join(tempRepo, "fixtures", "artifacts.json"),
      JSON.stringify(testCase.artifacts, null, 2),
    );
    await Bun.write(
      join(tempRepo, "fixtures", "policy.json"),
      JSON.stringify(testCase.policy, null, 2),
    );

    for (const gitCmd of [
      ["git", "init", "-q"],
      ["git", "config", "user.email", "harness@example.com"],
      ["git", "config", "user.name", "Act Harness"],
      ["git", "add", "-A"],
      ["git", "commit", "-qm", `act case ${testCase.name}`],
    ]) {
      const { exitCode, output } = run(gitCmd, tempRepo);
      if (exitCode !== 0) throw new Error(`${gitCmd.join(" ")} failed: ${output}`);
    }

    // 2. Run the actual workflow through act.
    console.log("  running `act push --rm` (this takes a while)...");
    // --pull=false: the runner image is local-only; act's default force-pull
    // would try (and fail) to fetch it from Docker Hub.
    const act = run(["act", "push", "--rm", "--pull=false"], tempRepo);

    // 3. Record the full output.
    resultLog +=
      `${"=".repeat(78)}\n=== ACT CASE: ${testCase.name} (exit code ${act.exitCode})\n` +
      `${"=".repeat(78)}\n${act.output}\n`;

    // 4. Assert on the outcome.
    const exp = testCase.expected;
    check(testCase.name, "act exit code is 0", act.exitCode === 0, `got ${act.exitCode}`);

    const succeeded = act.output.match(/Job succeeded/g) ?? [];
    checkEqual(testCase.name, "both jobs report 'Job succeeded'", succeeded.length, 2);
    check(testCase.name, "no job reports 'Job failed'", !/Job failed/.test(act.output));

    const planMatch = act.output.match(/::PLAN::(\{.*\})::ENDPLAN::/);
    check(testCase.name, "machine-readable plan found in CI logs", planMatch !== null);

    if (planMatch) {
      const plan = JSON.parse(planMatch[1]!);
      checkEqual(testCase.name, "summary matches exactly", plan.summary, exp.summary);
      checkEqual(testCase.name, "dryRun flag", plan.dryRun, exp.dryRun);
      checkEqual(
        testCase.name,
        "deleted artifact names",
        plan.toDelete.map((a: Artifact) => a.name).sort(),
        exp.deletedNames,
      );
      checkEqual(
        testCase.name,
        "retained artifact names",
        plan.toRetain.map((a: Artifact) => a.name).sort(),
        exp.retainedNames,
      );
      for (const [name, reasons] of Object.entries(exp.reasons)) {
        const doomed = plan.toDelete.find((a: Artifact) => a.name === name);
        checkEqual(testCase.name, `deletion reasons for ${name}`, doomed?.reasons, reasons);
      }
    }

    for (const needle of exp.stdoutContains) {
      check(testCase.name, `output contains ${JSON.stringify(needle)}`, act.output.includes(needle));
    }
    for (const needle of exp.stdoutNotContains) {
      // Only inspect the cleanup job's log lines: unit-test output inside the
      // "Run test suite" step legitimately mentions these strings.
      const cleanupLines = act.output
        .split("\n")
        .filter((l) => l.includes("Generate deletion plan"))
        .join("\n");
      check(
        testCase.name,
        `cleanup job output does NOT contain ${JSON.stringify(needle)}`,
        !cleanupLines.includes(needle),
      );
    }
  } finally {
    rmSync(tempRepo, { recursive: true, force: true });
  }
}

await Bun.write(RESULT_FILE, resultLog);
console.log(`\nact output written to ${RESULT_FILE}`);

if (failures.length > 0) {
  console.error(`\n${failures.length} assertion(s) FAILED:`);
  for (const f of failures) console.error(`  - ${f}`);
  process.exit(1);
}
console.log(`All ${CASES.length} act cases passed with exact expected values.`);
